### ISSUE TITLE
`google_monitoring_monitored_project` exponentially slow and spawns thousands of calls to cloudresourcemanager api

### ISSUE THREAD

**Issue by danjamesmay** [hashicorp/terraform-provider-google#12883]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

Terraform v1.3.3
on linux_amd64

### Affected Resource(s)

`google_monitoring_monitored_project`

### Terraform Configuration Files

```tf
locals {
    projects = jsondecode(file("projects.json"))
}

resource "google_monitoring_monitored_project" "project" {
    for_each =  toset(local.projects)

    metrics_scope = "my-scoping-project"
    name          = each.value
}
```

`projects.json` is a simple json array with project ids
```json
[
  "my-scoped-project"
]
```

### Debug Output

Snippet of output:
https://gist.github.com/danjamesmay/ce3e049831e6f6bb323964717d9f02b6

### Panic Output

n/a

### Expected Behavior

It appears the Google API supports full CRUD methods for interacting with monitored projects, via project ID or project number. The [resource code](https://github.com/hashicorp/terraform-provider-google/blob/main/google/resource_monitoring_monitored_project.go#L115) seems to be doing a lot more than just interacting with this API.

API:
- https://cloud.google.com/monitoring/api/ref_v3/rest/v1/locations.global.metricsScopes/get
- https://cloud.google.com/monitoring/api/ref_v3/rest/v1/locations.global.metricsScopes.projects/create
- https://cloud.google.com/monitoring/api/ref_v3/rest/v1/locations.global.metricsScopes.projects/delete

### Actual Behavior

Thousands upon thousands of calls to `cloudresourcemanager` resulting in rate limiting and very slow time to update state when reading and writing changes to the `google_monitoring_monitored_project` resource.

### Steps to Reproduce

When a project has 150+ projects listed, the time it takes for Terraform to check the state of each monitored project becomes slower and slower as rate limiting and exponential backoff limits Terraform's ability to work properly.

1. `terraform plan`

### Important Factoids

Nothing I can think of out of the ordinary.

### References

b/275113183

**Comment by danjamesmay** [comment]
I found a potentially useful line in the debug output:

```bash
2022-11-14T13:26:15.539Z [WARN]  Provider "registry.terraform.io/hashicorp/google" produced an invalid plan for google_monitoring_monitored_project.project["<redacted>"], but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .timeouts: planned for existence but config wants absence
```

**Comment by danjamesmay** [comment]
I set `parrelelism=1` in the plan options to serialise the projects.

Looks like it's doing a call to `GET /v1/locations/global/metricsScopes/<scoping_project>?alt=json HTTP/1.1` first, then iterating through every single item in the response to do a lookup to `GET /v1/projects/<for_each_key>?alt=json HTTP/1.1` for each key in `google_monitoring_monitored_project" "project"`.

It then seems to call `/v1/projects` over and over again while it performs the lookup below on the results from `/v1/locations/global/metricsSceope`. When it finds a match it continues.

2022-11-14T14:04:04.759Z [INFO]  provider.terraform-provider-google_v4.41.0_x5: 2022/11/14 14:04:04 [DEBUG][DCL INFO] Attempting to match <id from v1/projects call> with <id from /v1/locations/global/metricsSceopes>.: timestamp=2022-11-14T14:04:04.759Z

It seems to be a simple change to just store the results of each `/v1/projects` call in memory while it does the internal lookup, which would save all the thousands of calls and thus rate limiting.

The "time complexity" then increases for the amount of monitored projects as it has to loop through all of them in ascending order to match the individual project that Terraform is trying to refresh the state for. That then makes this really unwieldy when you have hundreds of projects.

**Comment by etruong42** [comment]
I found the issue. It appears to have been added June 2022. However, the trail is extremely difficult to follow because of all the ways the code gets autogenerated.

https://github.com/hashicorp/terraform-provider-google-beta/pull/4432/files#diff-4453ffdcd0c88d5bbf8e263267b9c0b6ddf437c2c4d39555c4ecaac20f8cd1dcR140

In this diff we can see "GetMonitoredProject" gets called, and within that definition, it can call CloudResourceManager for every monitored project.

I can't find the implementation of GetMonitoredProject, but we can see the same behavior in the current implementation: https://github.com/hashicorp/terraform-provider-google-beta/blob/main/google-beta/services/monitoring/resource_monitoring_monitored_project.go#L405
